lib/Basic/CMakeLists.txt

```diff
@@ -22,7 +27,7 @@ function(generate_revision_inc revision_inc_var name dir)
   add_custom_command(OUTPUT "${version_inc}"
     DEPENDS "${${name}_vc}" "${generate_vcs_version_script}"
     COMMAND ${CMAKE_COMMAND} "-DNAMES=$<UPPER_CASE:${name}>"
-                             "-D$<UPPER_CASE:${name}>_SOURCE_DIR=${dir}"
+                             "-D$<UPPER_CASE:${name}>_SOURCE_DIR=${swift_source_dir}"
```

Comment:
This path isn't specific to swift, why can't this remain using `dir`?

Reply:
I'm happy to rename it to e.g. source_dir. The point is to have a variable that is only defined when `SWIFT_APPEND_VC_REV` is true (`dir` is always defined in this place). `generate_vcs_version_script` hits different code path when `path` argument is falsey (https://github.com/llvm-mirror/llvm/blob/master/cmake/modules/GenerateVersionFromVCS.cmake#L21).

Reply:
Yeah, renaming the variable sounds like a good idea, alternatively, changing the value of the variable in the function is also possible since this is a CMake function rather than a macro.  The `swift_source_dir` is extremely confusing because `SWIFT_SOURCE_DIR` is the swift repository path.

When does the passed `dir` not exist?  I think that having some sort of comment around that is useful - I don't think that is expected.  This code was based on the LLVM flow, and it is possible that things have changed since and we should re-sync expectations.

Why can't we use `HAVE_VCS_VERSION_INC` and control that based on `SWIFT_APPEND_VC_REV` like the llvm projects?

Reply:
Renamed the variable to `dir_when_append_enabled`, wdyt?

> When does the passed `dir` not exist?

`dir` in `lib/Basic/CmakeLists.txt` always exists. If you ask when doesn't the `path` argument in the file `GenerateVersionFromVCS.cmake` exist, then in llvm repo empty `path` argument is passed when when `LLVM_APPEND_VC_REV` is disabled. I'd say this PR syncs the swift behavior with llvm's behavior.

I'm not sure I understand how `HAVE_VCS_VERSION_INC` works in llvm, it seems there are 2 knobs for the same thing, one setting `LLVM_APPEND_VC_REV` that results in generating `#undef`s only in a header, and second is setting `HAVE_VCS_VERSION_INC` and conditionally including that header. AFAIK having `SWIFT_APPEND_VC_REV` gets the better part of the llvm behavior. Plus we don't have a need for the `HAVE_VCS_VERSION_INC` in Swift (yet?).

Reply:
That variable seems overly verbose, but at least it's less confusing than `swift_source_dir`.

The `HAVE_VCS_VERSION_INC` indicates if the file is present or not.  That seems better in the sense that it's not relying on the truthiness of the variable.  It makes it more explicit that the header is not used, which seems clearer.  Is there something else that Im confusing here?

Reply:
> That variable seems overly verbose, but at least it's less confusing than swift_source_dir.

Could you suggest a better name? I can keep generating names until we find one that works best, but PR review turnaround is too slow for this approach :)

>The HAVE_VCS_VERSION_INC indicates if the file is present or not. That seems better in the sense that it's not relying on the truthiness of the variable. It makes it more explicit that the header is not used, which seems clearer. Is there something else that Im confusing here?

I'm positive it's me who doesn't understand things, I'm certainly not sure what exactly do you want me to do. Let me try to explain what I think you want. So in addition to `SWIFT_APPEND_VC_REV` you also want to set `HAVE_VCS_VERSION_INC` COMPILE_DEFINITIONS property on `lib/Basic/Version.cpp` (the only user of `SWIFT_REVISION`) and conditionally include the `SwiftRevision.inc` header based on `HAVE_VCS_VERSION_INC` when `SWIFT_APPEND_VC_REV` is truthy. I.e. `HAVE_VCS_VERSION_INC` and `SWIFT_APPEND_VC_REV` represent the same concept, just that former is a C++ preprocessor define, and the latter is Cmake variable. AFAIK that's the behaviour from llvm. I haven't done this yet because this step is IMHO unnecessary, but I don't feel strongly.

If I don't understand your point, would you mind explaining it from scratch once again? Thank you!

Reply:
I would say `location` is good, but by mirroring the LLVM pattern, the renaming wouldn't be needed at all right?

Yes, that's precisely it.  Rather than having the truthiness be based on the adjustment of the path, control the inclusion of the header via the macro like in LLVM.  It just makes it easier to have the projects be similar.  We can control the inclusion with the macro, which effectively treats the values as `undef` and so we don't have to play games with the path.

Reply:
Ok I took another look and I believe the LLVM pattern is as follows:

* `HAVE_VCS_VERSION_INC` is always defined in the Cmake build, independently from the `LLVM_APPEND_VC_REV`. E.g. https://github.com/llvm/llvm-project/blob/master/llvm/include/llvm/Support/CMakeLists.txt#L28 for llvm, the same applies to lld or lldb.
* the only effect `LLVM_APPEND_VC_REV` has is that the header with VCS information only contains `#undef`s. Passing empty path is the pattern used across llvm to disable putting VCS information into C++ code. This was introduced in https://github.com/llvm/llvm-project/commit/fb5fafb23cc2d8613f8be2487afe94d8594a88ce.

Therefore I believe this PR as is is precisely following the llvm pattern and doing what you suggested in the way I understand will deviate from the llvm pattern.

That said I agree that having `HAVE_VCS_VERSION_INC` defined conditionally depending on the value of `SWIFT_APPEND_VC_REV` is cleaner. I leave the decision what is more important to you, whether the consistency with llvm or cleaner implementation.

Reply:
Ah, that changed right after the branch point which explains a lot.

I do slightly prefer the cleaner approach for implementation.

(I am going to mark the change accepted as even though I prefer the cleaner implementation, this is still inline with llvm.  It would be nice if you did the cleaner implementation but I think that the fact that this is inline with LLVM means that this isn't any worse).

Reply:
Chatted with @gribozavr offline, and we ended up deciding to remove `HAVE_VCS_VERSION_INC` from llvm as well. We've decided that always generating a header is in sum less complex than conditionally generating a header. So we'll merge this PR as it.

Thank you so much for a thorough review (and for caring about the build!).